src/main/java/com/amazonaws/encryptionsdk/internal/AesGcmJceKeyCipher.java

```diff
@@ -34,54 +29,46 @@
     private static final int NONCE_LENGTH = 12;
     private static final int TAG_LENGTH = 128;
     private static final String TRANSFORMATION = "AES/GCM/NoPadding";
-    private final SecureRandom rnd = new SecureRandom();
 
     AesGcmJceKeyCipher(SecretKey key) {
         super(key, key);
     }
 
     private static byte[] specToBytes(final GCMParameterSpec spec) {
         final byte[] nonce = spec.getIV();
-        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        try (final DataOutputStream dos = new DataOutputStream(baos)) {
-            dos.writeInt(spec.getTLen());
-            dos.writeInt(nonce.length);
-            dos.write(nonce);
-            dos.close();
-            baos.close();
-        } catch (final IOException ex) {
-            throw new AssertionError("Impossible exception", ex);
-        }
-        return baos.toByteArray();
+        final byte[] result = new byte[Integer.BYTES + Integer.BYTES + nonce.length];
+        final ByteBuffer buffer = ByteBuffer.wrap(result);
+        buffer.putInt(spec.getTLen());
+        buffer.putInt(nonce.length);
+        buffer.put(nonce);
+        return result;
     }
 
     private static GCMParameterSpec bytesToSpec(final byte[] data, final int offset) throws InvalidKeyException {
-        final ByteArrayInputStream bais = new ByteArrayInputStream(data, offset, data.length - offset);
-        try (final DataInputStream dis = new DataInputStream(bais)) {
-            final int tagLen = dis.readInt();
-            final int nonceLen = dis.readInt();
+        final ByteBuffer buffer = ByteBuffer.wrap(data, offset, data.length - offset);
 
-            if (tagLen != TAG_LENGTH) {
-                throw new InvalidKeyException(String.format("Authentication tag length must be %s", TAG_LENGTH));
-            }
+        final int tagLen = buffer.getInt();
+        final int nonceLen = buffer.getInt();
```

Comment:
If the length is too short, you`BufferUnderflowException` which should be caught and rethrown as an InvalidKeyException.

Reply:
Defined a SPEC_LENGTH constant that I validate against